graviton/abi_strategy.py

```diff
@@ -183,3 +204,161 @@ def get(self) -> PyTypes:
         return cast(int, full_random) % (
             1 << (cast(abi.UintType, self.abi_type).bit_size // 2)
         )
+
+
+class ABIArgsMod(Enum):
+    selector_byte_insert = auto()
+    selector_byte_delete = auto()
+    selector_byte_replace = auto()
+    parameter_delete = auto()
+    parameter_append = auto()
+
+
+class ABICallStrategy:
+    """
+    TODO: refactor to comport with ABIStrategy + Hypothesis
+    TODO: make this generic on the strategy type
+    """
+
+    append_args_type: abi.ABIType = abi.ByteType()
+
+    def __init__(
+        self,
+        contract: str,
+        argument_strategy: Type[ABIStrategy] = RandomABIStrategy,
+        *,
+        num_dryruns: int = 1,
+        handle_selector: bool = True,
+        abi_args_mod: Optional[ABIArgsMod] = None,
+    ):
+        """
+        contract - ABI Contract JSON
+
+        argument_strategy (default=RandomABIStrategy) - ABI strategy for generating arguments
+
+        num_dry_runs (default=1) - the number of dry runs to run
+            (generates different inputs each time)
+
+        handle_selector (default=True) - usually we'll want to let
+            `ABIContractExecutor.run_sequence()`
+            handle adding the method selector so this param.
+            But if set False: when providing `inputs`
+            ensure that the 0'th argument for method calls is the selector.
+            And when set True: when NOT providing `inputs`, the selector arg
+            at index 0 will be added automatically.
+
+        abi_args_mod (optional) - when desiring to mutate the args, provide an ABIArgsMod value
+        """
+        self.contract: abi.Contract = abi.Contract.from_json(contract)
+        self.argument_strategy: Optional[Type[ABIStrategy]] = argument_strategy
```

Comment:
nit: I am a bit unconvinced about why `self.argument_strategy` should be optional, for we are assigning from argument `argument_strategy` of `Type[ABIStrategy]`.

Reply:
Good catch!

Reply:
cf. https://github.com/algorand/graviton/pull/49/commits/dfc2f7b0730006bc53f8dcd3013b92b7f8dd5052